data/tutorials/language/0lg_09_env.md

```diff
@@ -0,0 +1,610 @@
+```
+
+Inside the `max_42` function, the environment contains an additional binding between the first parameter of `max` and the value 42.
+
+## Recursive Functions
+
+In order to perform iterated computations, a function may call itself. Such a function is called _recursive_. In OCaml, recursive functions must be defined and explicitly declared by using `let rec`. It is not possible to accidentally create recursion loops between functions. As a consequence, recursive functions can't be anonymous.
+
+The classic (and very inefficient) way to present recursion is using the function that computes the [Fibonacci Sequence](https://en.wikipedia.org/wiki/Fibonacci_sequence).
+```ocaml
+# let rec fibo n = if n <= 1 then n else fibo (n - 1) + fibo (n - 2);;
+val fibo : int -> int = <fun>
+
+# List.init 10 Fun.id |> List.map fibo;;
+- : int list = [0; 1; 1; 2; 3; 5; 8; 13; 21; 34]
+```
+
+This version of `fibo` is inefficient because the number of recursive calls created doubles at each call, which creates exponential growth.
+
+This version does a better job:
+```ocaml
+# let rec fibo m n i = if i = 0 then m else fibo n (n + m) (i - 1);;
+val fibo : int -> int -> int -> int = <fun>
+
+# let fibo = fibo 0 1;;
+val fibo : int -> int = <fun>
+
+# List.init 10 Fun.id |> List.map fibo;;
+- : int list = [0; 1; 1; 2; 3; 5; 8; 13; 21; 34]
+```
+
+The first version takes two extra parameters: the two previously computed Fibonacci numbers.
+
+The second version uses the two first Fibonacci numbers as initial values. There is nothing to be computed when returning from a recursive call, so this enables the compiler to perform an optimisation called “tail call elimination.” This turns recursivity into imperative iteration in the generated native code and leads to much-improved performances.
+
+## Multiple Arguments Functions
+
+### Arrow Types and Mutiple Arguments
+
+Until now, functions with several arguments had a type looking like this: `a₁ -> a₂ -> a₃ -> ⋯ -> b`
+
+Where `a₁` is the type of the first argument, `a₂` is the type of the second argument, etc., and `b` is the type of the result. However, this type isn't displayed as it is. In reality, this type is the following: `a₁ -> (a₂ -> (a₃ -> ⋯ -> b))`. The arrow symbol is a [binary operator](https://en.wikipedia.org/wiki/Binary_operation). Here is how this can be observed:
+```ocaml
+# ( + );;
+- : int -> int -> int = <fun>
+
+# let f : (int -> (int -> int)) = ( + );;
+val f : int -> int -> int = <fun>
+```
+
+In this example, the type of integer addition function `( + )` is displayed first. It prints as `int -> int -> int`, meaning it takes two integers and returns an integer. In the second statement, `f` is defined to have type `int -> (int -> int)`. Despite the type of `( + )` looking different, the binding with `f` is allowed because those two types are the same. The toplevel always prints the version without parentheses to make it easier to read.
+
+Putting the parentheses the other way does not work:
```

Comment:
Why doesn't this work. Doesn't `int -> int -> int` mean it takes two integers and returns an integer? If so, why isn't it `(int -> int) -> int`, where the first `->` doesn't mean "returns." It seems the previous (li.457) example should give the error. (Maybe all these things don't need to be explained for your readers or in OCaml. It's okay if it's "because that's the way it is" and just needs to be learned that way in OCaml. Plus, I only know relatively basic maths, so it might be clear to your readers.) ... Still, I want to raise the question just in case!